ironfish-cli/src/commands/service/snapshot.ts

```diff
@@ -0,0 +1,213 @@
+
+    progress.start(stop - start + 1, 0)
+
+    for await (const result of response.contentStream()) {
+      if (result.buffer && result.seq) {
+        const blockFilePath = this.sdk.fileSystem.join(blockExportPath, `${result.seq}`)
+        await fsAsync.writeFile(blockFilePath, Buffer.from(result.buffer))
+        progress.update(result.seq || 0)
+      }
+    }
+
+    progress.stop()
+
+    const timestamp = Date.now()
+
+    const snapshotFileName = `ironfish_snapshot_${timestamp}.tar.gz`
+    const snapshotPath = this.sdk.fileSystem.join(exportDir, snapshotFileName)
+
+    this.log(`Zipping\n    SRC ${blockExportPath}\n    DST ${snapshotPath}\n`)
+    CliUx.ux.action.start(`Zipping ${blockExportPath}`)
+    await this.zipDir(blockExportPath, snapshotPath)
+    const stat = await fsAsync.stat(snapshotPath)
+    const fileSize = stat.size
+    CliUx.ux.action.stop(`done (${FileUtils.formatFileSize(fileSize)})`)
+
+    const hasher = crypto.createHash('sha256')
+    const fileHandle = await fsAsync.open(snapshotPath, 'r')
+    const stream = fileHandle.createReadStream()
+
+    CliUx.ux.action.start(`Creating checksum for ${snapshotPath}`)
+    for await (const data of stream) {
+      hasher.update(data)
+    }
+    const checksum = hasher.digest().toString('hex')
+    CliUx.ux.action.stop(`done`)
+
+    const blockHeight = stop
+
+    CliUx.ux.action.start(`Uploading to ${bucket}`)
+    await this.uploadToBucket(snapshotPath, bucket, 'application/x-compressed-tar')
+    CliUx.ux.action.stop(`done`)
+
+    await fsAsync.writeFile(
+      path.join(exportDir, 'manifest.json'),
+      JSON.stringify({
+        block_height: blockHeight,
+        checksum,
+        file_name: snapshotFileName,
+        file_size: fileSize,
+        timestamp,
+      }),
+    )
+
+    CliUx.ux.action.start(`Uploading latest snapshot information to ${bucket}`)
+    await this.uploadToBucket(path.join(exportDir, 'manifest.json'), bucket, 'application/json')
+    CliUx.ux.action.stop(`done`)
+  }
+
+  zipDir(source: string, dest: string, excludes: string[] = []): Promise<number | null> {
+    return new Promise<number | null>((resolve, reject) => {
+      const sourceDir = path.dirname(source)
+      const sourceFile = path.basename(source)
+
+      const args = ['-zcf', dest, '-C', sourceDir, sourceFile]
+
+      for (const exclude of excludes) {
+        args.unshift(exclude)
+        args.unshift('--exclude')
+      }
+
+      const process = spawn('tar', args)
+      process.on('exit', (code) => resolve(code))
+      process.on('close', (code) => resolve(code))
+      process.on('error', (error) => reject(error))
+    })
+  }
+
+  uploadToBucket(dest: string, host: string, contentType: string): Promise<number | null> {
+    return new Promise<number | null>((resolve, reject) => {
+      const date = new Date().toISOString()
+      const file = path.basename(dest)
+      const acl = 'bucket-owner-full-control'
+
+      const process = spawn(
+        `curl`,
```

Comment:
Invoking a command line process from within code is generally an antipattern. We use `axios` for this kind of thing elsewhere, would recommend that. Or more directly you can use `node-libcurl` if axios doesn't work (but it should!)

Reply:
actually, even better than both of these is the AWS node SDK

https://stackabuse.com/uploading-files-to-aws-s3-with-node-js/

Reply:
If we're moving away from hardcoding a `curl`call that includes AWS-specific data, is it prudent to use an SDK that ties users to that service directly?

Reply:
Well, the parameter for the CLI command is `bucket` and it seems every assumption is made that this is an AWS bucket, so I would think using the AWS SDK is fine.

And to clarify - my comment about the curl command wasn't due to AWS tie-in, it was because you generally shouldn't be invoking OS-level commands from code.

Reply:
Sure, I can use the SDK there.

Agreed on the second, just opening up for alternatives there.

Reply:
@lwisne: After thinking about it and tinkering with some implementations, I'm not sure that moving to an SDK is the best move here. If we include it as part of the command, then every user who pulls down our software is going to have to download the entire AWS SDK and all of its dependencies. Furthermore, I think it forces users of this command to use AWS and that may not be a choice they're willing to make.

Additionally, for our use case, AWS requires multiple requests to be made to open a multipart file upload session as they now require it for large files. If we keep `curl` or even move to alternatives (e.g. `node-libcurl`, `axios`) for uploading to S3, hard-coding AWS' multiple request requirement may prove cumbersome for other users who may want to use a different host.

As a alternative option, @NullSoldier mentioned that we could leverage GitHub actions to upload the file(s) and its manifest to AWS, if we so choose. That would reduce some of the complexity of the command (allowing for easier unit testing) and it would allow us to potentially move this command to `chain:export` instead, which would address your concerns with it being hidden.

Reply:
Can you explain the Github action methodology a bit more? Not familiar

Reply:
We can define a workflow to deploy artifacts to S3. See [this page](https://dev.to/johnkevinlosito/deploy-static-website-to-s3-using-github-actions-4a0e) for an example. I'm envisioning an AWS instance exporting a snapshot and returning the path of the snapshot, and the GitHub action instructing the instance to send it to S3 through use of the AWS SDK. It doesn't even have to be a GitHub action, really; it could be a separate script that leverages the SDK.

In any case, I'm not sure that keeping the upload operation as a part of the command is the best idea. The most user-facing part of this project is arguably the snapshot import; many more people will use that than the export command. Reducing the scope of the export command to just _exporting_ allows for more choice in what to do with the resultant file; one could decide to just use this command as a backup solution or use it to get an archive that can place on a non-AWS host, e.g. GCP, Yandex, etc.

Reply:
This is definitely something we can do, but IMO these options represent scope creep/overengineering vs the original spec. You had export uploading to a S3 bucket, and only an S3 bucket, and I think sticking to that will be a perfectly solid way to go about it for v1.

And if we are going to upload to an S3 bucket, using Amazon's well-vetted tools to do so seems to be the most robust route. Jason told me the bit of cURL code in `backup.ts` is broken going forward, because the file is too big and needs to be handled in some other fashion in order to do a S3 upload. This is exactly the kind of issue that using the SDK would likely prevent, in addition to my current comment about handling error states gracefully.

I'm just having a lot of trouble seeing the problem we are avoiding here by not choosing the simple, obvious solution of uploading a file to S3 via Amazon's S3 SDK.

Reply:
Fair enough, I'll use the SDK then.